In [2]:
import pandas as pd

In [5]:
file_path="../data/raw/online_retail_II.xlsx"
excel_file=pd.ExcelFile(file_path)
excel_file.sheet_names

['Year 2009-2010', 'Year 2010-2011']

In [6]:
df_2009_2010=pd.read_excel(file_path,sheet_name=excel_file.sheet_names[0])
df_2010_2011=pd.read_excel(file_path,sheet_name=excel_file.sheet_names[1])

In [8]:
print("size of 2009 - 2010 sheet : ",df_2009_2010.shape)
print("size of 2010 - 2011 sheet : ",df_2010_2011.shape)

size of 2009 - 2010 sheet :  (525461, 8)
size of 2010 - 2011 sheet :  (541910, 8)


In [9]:
print("Columns present in 2009 - 2010 sheet : ",df_2009_2010.columns.tolist())
print("Columns present in 2010 - 2011 sheet : ",df_2010_2011.columns.tolist())

Columns present in 2009 - 2010 sheet :  ['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']
Columns present in 2010 - 2011 sheet :  ['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']


In [10]:
df_2009_2010.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[ns]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


In [11]:
df_2010_2011.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[ns]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [12]:
df_2009_2010.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,525461.000000,525461,525461.000000,417534.000000
mean,10.337667,2010-06-28 11:37:36.845017856,4.688834,15360.645478
min,-9600.000000,2009-12-01 07:45:00,-53594.360000,12346.000000
25%,1.000000,2010-03-21 12:20:00,1.250000,13983.000000
50%,3.000000,2010-07-06 09:51:00,2.100000,15311.000000
75%,10.000000,2010-10-15 12:45:00,4.210000,16799.000000
max,19152.000000,2010-12-09 20:01:00,25111.090000,18287.000000
std,107.424110,NaN,146.126914,1680.811316


In [13]:
df_2010_2011.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,541910.000000,541910,541910.000000,406830.000000
mean,9.552234,2011-07-04 13:35:22.342307584,4.611138,15287.684160
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.080957,NaN,96.759765,1713.603074


In [14]:
df_2009_2010.isna().sum()

Invoice             0
StockCode           0
Description      2928
Quantity            0
InvoiceDate         0
Price               0
Customer ID    107927
Country             0
dtype: int64

In [15]:
missing_2009_2010=(df_2009_2010.isna().mean()*100).sort_values(ascending=False)
missing_2009_2010

Customer ID    20.539488
Description     0.557225
StockCode       0.000000
Invoice         0.000000
Quantity        0.000000
InvoiceDate     0.000000
Price           0.000000
Country         0.000000
dtype: float64

In [16]:
df_2010_2011.isna().sum()

Invoice             0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
Price               0
Customer ID    135080
Country             0
dtype: int64

In [17]:
missing_2010_2011=(df_2010_2011.isna().mean()*100).sort_values(ascending=False)
missing_2010_2011

Customer ID    24.926648
Description     0.268310
StockCode       0.000000
Invoice         0.000000
Quantity        0.000000
InvoiceDate     0.000000
Price           0.000000
Country         0.000000
dtype: float64

In [18]:
df_2009_2010.nunique()

Invoice        28816
StockCode       4632
Description     4681
Quantity         825
InvoiceDate    25296
Price           1606
Customer ID     4383
Country           40
dtype: int64

In [19]:
df_2010_2011.nunique()

Invoice        25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
Price           1630
Customer ID     4372
Country           38
dtype: int64

In [20]:
print("Minimum and Maximum of the 2009 - 2010 sheet for invoiceDate :")
print(df_2009_2010["InvoiceDate"].min())
print(df_2009_2010["InvoiceDate"].max())

Minimum and Maximum of the 2009 - 2010 sheet for invoiceDate :
2009-12-01 07:45:00
2010-12-09 20:01:00


In [21]:
print("Minimum and Maximum of the 2010 - 2011 sheet for invoiceDate :")
print(df_2010_2011["InvoiceDate"].min())
print(df_2010_2011["InvoiceDate"].max())

Minimum and Maximum of the 2010 - 2011 sheet for invoiceDate :
2010-12-01 08:26:00
2011-12-09 12:50:00


## Transaction Behaviour Analysis

In [22]:
is_cancelled_09_10 = df_2009_2010["Invoice"].astype("str").str.startswith("C")
normal_tx_09_10=df_2009_2010[~is_cancelled_09_10]
cancel_tx_09_10=df_2009_2010[is_cancelled_09_10]
print("Count of Normal Transactions : ",len(normal_tx_09_10))
print("Count of Cancelled Transactions : ",len(cancel_tx_09_10))
print("Rate of cancellation : ",f"{is_cancelled_09_10.mean()*100 :.2f}%")

Count of Normal Transactions :  515255
Count of Cancelled Transactions :  10206
Rate of cancellation :  1.94%


In [23]:
is_cancelled_10_11 = df_2010_2011["Invoice"].astype("str").str.startswith("C")
normal_tx_10_11 = df_2010_2011[~is_cancelled_10_11]
cancel_tx_10_11 = df_2010_2011[is_cancelled_10_11]
print("Count of Normal Transactions : ",len(normal_tx_10_11))
print("Count of Cancelled Transactions : ",len(cancel_tx_10_11))
print("Rate of cancellation : ",f"{is_cancelled_10_11.mean()*100:.2f}%")

Count of Normal Transactions :  532622
Count of Cancelled Transactions :  9288
Rate of cancellation :  1.71%


In [24]:
negative_quantity = df_2009_2010[df_2009_2010["Quantity"]<0]
negative_quantity.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia


In [25]:
negative_quantity["Invoice"].astype("str").str.startswith("C").value_counts()


Invoice
True     10205
False     2121
Name: count, dtype: int64

In [26]:
negative_price=df_2009_2010[df_2009_2010["Price"]<0]
negative_price.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom


In [27]:
invoice_counts = df_2009_2010["Invoice"].value_counts()
invoice_counts.head()

Invoice
537434    675
538071    652
537638    601
537237    597
536876    593
Name: count, dtype: int64

In [28]:
example_invoice = invoice_counts.index[0]
df_2009_2010[df_2009_2010["Invoice"]==example_invoice]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
516208,537434,20685,DOORMAT RED RETROSPOT,1,2010-12-06 16:57:00,14.43,NaN,United Kingdom
516209,537434,20699,MOUSEY LONG LEGS SOFT TOY,1,2010-12-06 16:57:00,5.06,NaN,United Kingdom
516210,537434,20713,JUMBO BAG OWLS,1,2010-12-06 16:57:00,4.21,NaN,United Kingdom
516211,537434,20719,WOODLAND CHARLOTTE BAG,3,2010-12-06 16:57:00,1.66,NaN,United Kingdom
516212,537434,20725,LUNCH BAG RED RETROSPOT,4,2010-12-06 16:57:00,4.21,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
516878,537434,20652,BLUE POLKADOT LUGGAGE TAG,2,2010-12-06 16:57:00,2.51,NaN,United Kingdom
516879,537434,20658,RED RETROSPOT LUGGAGE TAG,1,2010-12-06 16:57:00,2.51,NaN,United Kingdom
516880,537434,20665,RED RETROSPOT PURSE,1,2010-12-06 16:57:00,5.91,NaN,United Kingdom
516881,537434,20682,RED RETROSPOT CHILDRENS UMBRELLA,1,2010-12-06 16:57:00,6.77,NaN,United Kingdom


In [29]:
customer_purchase_days = (df_2009_2010.dropna(subset=["Customer ID"]).groupby("Customer ID")["InvoiceDate"].nunique())
customer_purchase_days.describe()

count    4383.000000
mean        5.359115
std         9.958893
min         1.000000
25%         1.000000
50%         3.000000
75%         6.000000
max       267.000000
Name: InvoiceDate, dtype: float64

In [30]:
customer_orders = (df_2009_2010.dropna(subset=["Customer ID"]).groupby("Customer ID")["Invoice"].nunique())
customer_orders.describe()

count    4383.000000
mean        5.381474
std        10.051921
min         1.000000
25%         1.000000
50%         3.000000
75%         6.000000
max       270.000000
Name: Invoice, dtype: float64

## Customer Purchase Behaviour Analysis

In [31]:
df = pd.concat([df_2009_2010,df_2010_2011],ignore_index=True)
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [32]:
customer_df = df.dropna(subset=["Customer ID"]).copy()
customer_df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [33]:
customer_df["Invoice"] = customer_df["Invoice"].astype("str")
purchase_df = customer_df[~customer_df["Invoice"].str.startswith("C")].copy()
purchase_df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [34]:
customer_dates = purchase_df[["Customer ID","Invoice","InvoiceDate"]].drop_duplicates(["Customer ID","Invoice"]).sort_values(["Customer ID","InvoiceDate"])
customer_dates.head()

,Customer ID,Invoice,InvoiceDate
27994,12346.0,491725,2009-12-14 08:34:00
28251,12346.0,491742,2009-12-14 11:00:00
28254,12346.0,491744,2009-12-14 11:02:00
39398,12346.0,492718,2009-12-18 10:47:00
39411,12346.0,492722,2009-12-18 10:55:00


In [35]:
customer_dates["previous_purchase"] = customer_dates.groupby("Customer ID")["InvoiceDate"].shift(1)
customer_dates.head()

,Customer ID,Invoice,InvoiceDate,previous_purchase
27994,12346.0,491725,2009-12-14 08:34:00,NaT
28251,12346.0,491742,2009-12-14 11:00:00,2009-12-14 08:34:00
28254,12346.0,491744,2009-12-14 11:02:00,2009-12-14 11:00:00
39398,12346.0,492718,2009-12-18 10:47:00,2009-12-14 11:02:00
39411,12346.0,492722,2009-12-18 10:55:00,2009-12-18 10:47:00


In [36]:
customer_dates["purchase_interval_days"] = (customer_dates["InvoiceDate"]-customer_dates["previous_purchase"]).dt.total_seconds()/(24*60*60)
customer_dates.head()

,Customer ID,Invoice,InvoiceDate,previous_purchase,purchase_interval_days
27994,12346.0,491725,2009-12-14 08:34:00,NaT,NaN
28251,12346.0,491742,2009-12-14 11:00:00,2009-12-14 08:34:00,0.101389
28254,12346.0,491744,2009-12-14 11:02:00,2009-12-14 11:00:00,0.001389
39398,12346.0,492718,2009-12-18 10:47:00,2009-12-14 11:02:00,3.989583
39411,12346.0,492722,2009-12-18 10:55:00,2009-12-18 10:47:00,0.005556


In [37]:
customer_dates["purchase_interval_days"].describe()

count    31094.000000
mean        51.681669
std         75.786619
min          0.000000
25%          6.983333
50%         24.740972
75%         61.804340
max        714.152083
Name: purchase_interval_days, dtype: float64

In [38]:
customer_interval_stats = (customer_dates.groupby("Customer ID")["purchase_interval_days"].agg(mean_interval="mean",median_interval="median",max_interval="max",std_interval="std",purchase_count="count"))
customer_interval_stats.describe()

,mean_interval,median_interval,max_interval,std_interval,purchase_count
count,4255.000000,4255.000000,4255.000000,3311.000000,5881.000000
mean,102.298261,92.712161,174.933083,69.278077,5.287196
std,98.277802,101.178308,122.637918,61.785159,13.012879
min,0.000000,0.000000,0.000000,0.000491,0.000000
25%,38.576845,28.079861,80.063194,26.775430,0.000000
50%,72.331019,55.972222,151.828472,52.138934,2.000000
75%,129.843750,118.185764,244.980556,90.415635,6.000000
max,714.152083,714.152083,714.152083,495.748636,397.000000


In [39]:
customer_interval_stats["median_interval"].quantile([0.5,0.75,0.9,0.95,0.99])

0.50     55.972222
0.75    118.185764
0.90    220.392986
0.95    322.161181
0.99    484.492944
Name: median_interval, dtype: float64

In [40]:
customer_interval_stats["max_interval"].quantile([0.5,0.75,0.9,0.95,0.99])

0.50    151.828472
0.75    244.980556
0.90    352.237639
0.95    391.151111
0.99    540.499542
Name: max_interval, dtype: float64

In [41]:
(customer_interval_stats["purchase_count"] == 0).sum()

np.int64(1626)

In [42]:
customer_order_counts = (purchase_df.drop_duplicates(["Customer ID","Invoice"]).groupby("Customer ID").size())
customer_order_counts.describe()

count    5881.000000
mean        6.287196
std        13.012879
min         1.000000
25%         1.000000
50%         3.000000
75%         7.000000
max       398.000000
dtype: float64

In [43]:
(customer_order_counts == 1).sum()

np.int64(1626)

In [46]:
customer_last_purchase = purchase_df.groupby("Customer ID")["InvoiceDate"].max()
dataset_end = purchase_df["InvoiceDate"].max()

customer_recency = (dataset_end - customer_last_purchase).dt.total_seconds()/(24*60*60)

In [47]:
customer_recency.describe()

count    5881.000000
mean      200.980188
std       209.486712
min         0.000000
25%        25.053472
50%        95.040278
75%       379.110417
max       738.121528
Name: InvoiceDate, dtype: float64

In [48]:
customer_recency.quantile([0.25,0.5,0.75,0.9,0.95,0.99])

0.25     25.053472
0.50     95.040278
0.75    379.110417
0.90    535.091667
0.95    625.004861
0.99    725.871944
Name: InvoiceDate, dtype: float64

In [50]:
thresholds=[30,60,90,120,180]
for threshold in thresholds:
    count = (customer_recency>=threshold).sum()
    percentage = (count/len(customer_recency))*100
    print(f"{threshold} days : {count} customers({percentage:.2f}%)")

30 days : 4232 customers(71.96%)
60 days : 3484 customers(59.24%)
90 days : 2991 customers(50.86%)
120 days : 2759 customers(46.91%)
180 days : 2402 customers(40.84%)


In [51]:
(df["Quantity"] == 0).sum()

np.int64(0)

In [52]:
df.duplicated().sum()

np.int64(34335)

In [53]:
df[df["Price"]<0]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom
825444,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
825445,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [58]:
negative_normal = df[(df["Quantity"]<0 ) & (~df["Invoice"].astype("str").str.startswith("C"))]
negative_normal.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,NaN,United Kingdom
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom
4538,489820,21133,invcd as 84879?,-720,2009-12-02 13:23:00,0.0,NaN,United Kingdom
4566,489821,85049G,NaN,-240,2009-12-02 13:25:00,0.0,NaN,United Kingdom


In [59]:
duplicates = df[df.duplicated(keep=False)].copy()
duplicates.shape

(67242, 8)

In [60]:
duplicates.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
367,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom


In [61]:
duplicates["Invoice"].nunique()

5391

In [64]:
duplicates.groupby("Invoice").size().describe()

count    5391.000000
mean       12.473011
std        55.443206
min         2.000000
25%         2.000000
50%         4.000000
75%         8.000000
max      1350.000000
dtype: float64

In [63]:
duplicate_invoice_counts = (
    duplicates
    .groupby("Invoice")
    .size()
    .sort_values(ascending=False)
)

duplicate_invoice_counts.head(20)

Invoice
537434    1350
538071    1304
537638    1202
537237    1194
536876    1186
536592    1184
537823    1182
537240    1136
537666    1072
536544    1054
537224     338
537781     322
537640     298
536982     290
536865     280
537643     246
537642     244
538153     242
536796     242
537126     236
dtype: int64

In [65]:
example_invoice = duplicate_invoice_counts.index[0]

df[df["Invoice"] == example_invoice]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
516208,537434,20685,DOORMAT RED RETROSPOT,1,2010-12-06 16:57:00,14.43,NaN,United Kingdom
516209,537434,20699,MOUSEY LONG LEGS SOFT TOY,1,2010-12-06 16:57:00,5.06,NaN,United Kingdom
516210,537434,20713,JUMBO BAG OWLS,1,2010-12-06 16:57:00,4.21,NaN,United Kingdom
516211,537434,20719,WOODLAND CHARLOTTE BAG,3,2010-12-06 16:57:00,1.66,NaN,United Kingdom
516212,537434,20725,LUNCH BAG RED RETROSPOT,4,2010-12-06 16:57:00,4.21,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
539401,537434,20652,BLUE POLKADOT LUGGAGE TAG,2,2010-12-06 16:57:00,2.51,NaN,United Kingdom
539402,537434,20658,RED RETROSPOT LUGGAGE TAG,1,2010-12-06 16:57:00,2.51,NaN,United Kingdom
539403,537434,20665,RED RETROSPOT PURSE,1,2010-12-06 16:57:00,5.91,NaN,United Kingdom
539404,537434,20682,RED RETROSPOT CHILDRENS UMBRELLA,1,2010-12-06 16:57:00,6.77,NaN,United Kingdom


In [66]:
duplicate_rows = df[
    df.duplicated(keep=False)
].sort_values(
    ["Invoice", "StockCode", "Description", "Quantity", "Price"]
)

duplicate_rows.head(30)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom


In [67]:
example = duplicate_rows.iloc[0]

df[
    (df["Invoice"] == example["Invoice"]) &
    (df["StockCode"] == example["StockCode"]) &
    (df["Quantity"] == example["Quantity"]) &
    (df["Price"] == example["Price"])
]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom


In [68]:
duplicate_groups = (
    df.groupby(
        ["Invoice", "StockCode", "Description",
         "Quantity", "InvoiceDate", "Price",
         "Customer ID", "Country"],
        dropna=False
    )
    .size()
    .sort_values(ascending=False)
)

duplicate_groups.head(20)

Invoice  StockCode  Description                          Quantity  InvoiceDate          Price  Customer ID  Country       
555524   22698      PINK REGENCY TEACUP AND SAUCER       1         2011-06-05 11:37:00  2.95   16923.0      United Kingdom    20
         22697      GREEN REGENCY TEACUP AND SAUCER      1         2011-06-05 11:37:00  2.95   16923.0      United Kingdom    12
537224   70007      HI TEC ALPINE HAND WARMER            1         2010-12-05 16:24:00  1.65   13174.0      United Kingdom    10
572861   22775      PURPLE DRAWERKNOB ACRYLIC EDWARDIAN  12        2011-10-26 12:46:00  1.25   14102.0      United Kingdom     8
536796   21967      PACK OF 12 SKULL TISSUES             1         2010-12-02 15:46:00  0.29   15574.0      United Kingdom     6
572344   M          Manual                               48        2011-10-24 10:43:00  1.50   14607.0      United Kingdom     6
536874   22866      HAND WARMER SCOTTY DOG DESIGN        1         2010-12-03 11:35:00  2.10   16891.0 

## Intial Data Observation
1. Number of unique customers are less than total transaction records indicating one customer has several transaction records
2. Around 21% of customer_ids are missing in 2009 - 2010 and 26% of customer_ids are missing in 2010 - 2011 transaction sheets
3. Negative quantities occur in the transaction data and appear to represent cancellations/returns, while negative unit prices also exist and require      separate investigation.
4. Value in Price field is unit price instead of whole transaction price
5. The number of observed customer IDs differs between the two sheets, but this difference alone cannot establish customer churn. Customer retention needs to be analyzed using customer activity over time and an explicit churn definition.
6. Approximately 1.94% of transaction-line records have cancellation-style invoice identifiers, while approximately 98.06% are normal invoice records.
7. negative quantity rows most(10205) of them are having invoice number starting with C while other (2121) rows are starting with normal invoice but with missing customer id
8. negative price rows are only 3 in 2009 - 2010 and all of them having invoice with starting A and stock code as B
9. one invoice have multiple rows indicating one invoice contain multiple product orders
10. customer purchase days total unique are 4383 with minimum 1 and maximum 267 and customer order total unique are 4383 with minimum 1 and maximum 270
11. The median purchase interval is approximately 24.7 days, while the mean is 51.7 days, indicating a strongly right-skewed distribution.
12. Approximately 75% of observed purchase intervals are below 62 days, suggesting that extended inactivity beyond this range may represent meaningful behavioral change.


## Proposed Churn Defination : 

1. Oberservation Window : 180 Days
2. Prediction Window : 90 Days
3. Churn Defination : A customer is considered behaviorally churned if they make no qualifying purchase during the 90-day prediction window.

### Rationale:
1. The customer-level median purchase interval is approximately 56 days, while the 75th percentile is approximately 118 days. A 180-day observation window therefore provides sufficient historical context to capture multiple purchasing cycles for many customers and allows us to engineer meaningful behavioral features such as recency, frequency, monetary value, purchase intervals, and product diversity.
2. A 90-day prediction window provides a sufficiently long future period to identify meaningful customer inactivity while remaining operationally useful for retention planning.
3. Prediction snapshots will only be created when the complete 90-day future period exists in the dataset. This prevents right-censored observations from being incorrectly labeled as churn.